### PR TITLE
CDPT-1475: Intranet migration - content freeze

### PR DESCRIPTION
**Pull request by EmilyHazlehurst**
- Add maintenance class
- Add settings page for maintenance functions
- Add functionality to enable maintenance mode and lock non-admins out of the admin section of the site
- Add functionality to display a notification banner as advanced warning for editors of upcoming maintenance

### CODE REVIEW

File: web/app/themes/clarity/inc/maintenance.php [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+     *
+     * @param array $args A mixed array of arguments to render the different field types
+     */
+    public function render_field(array $args): void
+    {
+
+        $field = $args['field'];
+
+        // Get the value of the setting we've registered with register_setting()
+        $options = get_option('maintenance_options');
```
Comment (wilson1000) [inline web/app/themes/clarity/inc/maintenance.php:218]: Hey @EmilyHazlehurst - on the first test of the class I received a fatal error. This option array will only exist once values have been submitted. 

It is however possible to use the 2nd argument of the `get_option()` function to define a default array, for example:

```php
get_option('maintenance_options', [
    'notification_content' => '',
    'notification_title' => ''
]);
```

Comment (wilson1000) [reply web/app/themes/clarity/inc/maintenance.php:218]: https://developer.wordpress.org/reference/functions/get_option/

Comment (wilson1000) [reply web/app/themes/clarity/inc/maintenance.php:218]:
```php
Warning: Trying to access array offset on value of type bool in /bedrock/web/app/themes/clarity/inc/maintenance.php on line 220
```

File: web/app/themes/clarity/inc/maintenance.php [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+    }
+
+    /**
+     * Generates a unique id based on the content of the banner
+     *
+     * @param string $content The banner's content
+     * @param string $title The banner's title
+     * @return string
+     */
+    private function get_notification_hash(string $content, string $title): string
```
Comment (wilson1000) [inline web/app/themes/clarity/inc/maintenance.php:392]:
```php
Fatal error: Uncaught TypeError: MOJ\Intranet\Maintenance::get_notification_hash(): Argument #1 ($content) must be of type string, null given, called in /bedrock/web/app/themes/clarity/inc/maintenance.php on line 220 and defined in /bedrock/web/app/themes/clarity/inc/maintenance.php:392 Stack trace: #0 /bedrock/web/app/themes/clarity/inc/maintenance.php(220): MOJ\Intranet\Maintenance->get_notification_hash(NULL, NULL) #1 /bedrock/web/wp/wp-admin/includes/template.php(1836): MOJ\Intranet\Maintenance->render_field(Array) #2 /bedrock/web/wp/wp-admin/includes/template.php(1790): do_settings_fields('maintenance', 'maintenance-sec...') #3 /bedrock/web/app/themes/clarity/inc/maintenance.php(194): do_settings_sections('maintenance') #4 /bedrock/web/wp/wp-includes/class-wp-hook.php(324): MOJ\Intranet\Maintenance->render_options_page('') #5 /bedrock/web/wp/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters('', Array) #6 /bedrock/web/wp/wp-includes/plugin.php(517): WP_Hook->do_action(Array) #7 /bedrock/web/wp/wp-admin/admin.php(259): do_action('settings_page_m...') #8 /bedrock/web/wp/wp-admin/options-general.php(10): require_once('/bedrock/web/wp...') #9 {main} thrown in /bedrock/web/app/themes/clarity/inc/maintenance.php on line 392
```

File: web/app/themes/clarity/inc/maintenance.php [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+            {
+                ?>
+                <input
+                    type="text"
+                    id="<?php echo esc_attr($field['id']); ?>"
+                    name="maintenance_options[<?php echo esc_attr($field['id']); ?>]"
+                    value="<?php echo isset($options[$field['id']]) ? esc_attr($options[$field['id']]) : ''; ?>"
+                >
+                <p class="description">
+                    <?php esc_html_e($field['description'], 'maintenance'); ?>
```
Comment (wilson1000) [inline web/app/themes/clarity/inc/maintenance.php:233]:
```php
Warning: Undefined array key "description" in /bedrock/web/app/themes/clarity/inc/maintenance.php on line 237
```

**Review by wilson1000 — CHANGES_REQUESTED** [review]
Hi Emily, I've managed to solve the errors locally by defining default values for array keys that may not exist.

**Incidentally**
There is a dismissible notification popup available in the MoJ Components plugin. Much like the one you have created here.
https://intranet2.dev.wp.dsd.io/wp/wp-admin/options-general.php?page=mojComponentSettings&moj-tab=component-tab-0 

This also features user specific dismissals and upon save, it shows the new message to users by removing all transients previously stored when a dismissal was logged.

File: web/app/themes/clarity/inc/maintenance.php (hunk 3 above)
Comment (EmilyHazlehurst) [reply web/app/themes/clarity/inc/maintenance.php]: This has been updated to only access the description array key if a value is set.

File: web/app/themes/clarity/inc/maintenance.php (hunk 1 above)
Comment (EmilyHazlehurst) [reply web/app/themes/clarity/inc/maintenance.php]: Thanks @wilson1000, I've updated this to have default values.

**Review by wilson1000 — APPROVED** [review]
Looks good 😄 nice work!

**Pull request merged**